### DISCUSSION TITLE
How to render UserControl on Android when using a secondary display?

### DISCUSSION THREAD

**Discussion by DoubleDBE** [AvaloniaUI/Avalonia#17298] [Q&A] [answered]
In my application, I need to render a UserControl onto the secondary display of my Android device.

On desktop it's very easy to do because I can use a Window, but for Android I am at a loss:

Am I missing something stupid?

```
var topLevel = GetTopLevel(this);

var screen = topLevel.Screens.All.FirstOrDefault(x => x.IsPrimary is false);

if (screen is not null)
{
    Position = screen.WorkingArea.Position;
}

Content = new myUserControl();
```

In the Xamarin.Forms days, I have done this before where I needed to subclass the `Presentation` class and get the actual android view from XF's renderer, ie;

```
class MyCustomPresentation : Presentation
{
    public MyCustomPresentation(Context outerContext, Display display) : base(outerContext, display)
    {
    }

    protected override void OnCreate(Bundle savedInstanceState)
    {
        base.OnCreate(savedInstanceState);

        // set content here
    }
}
```

And show it like this:

```
DisplayManager displayManager = (DisplayManager)context.GetSystemService(Context.DisplayService);

Display[] presentationDisplays = displayManager.GetDisplays(DisplayManager.DisplayCategoryPresentation);

if (presentationDisplays?.Length > 0)
{
    Display display = presentationDisplays[0];

    var myCustomPresentation = new MyCustomPresentation(context, display);

    myCustomPresentation.Show();
}
```

**Comment by maxkatz6** [comment]
Mobile Avalonia doesn't have APIs for windows yet.
You can use the same `Presentation` code from your Android specific project, for example, or via an interface abstraction. If necessary, you can create a custom `AvaloniaView` (as an Android control) inside of that presentation, if you need to render Avalonia inside of custom activities/presentations. But we don't have samples for that.

**Reply by DoubleDBE** [reply]
> Mobile Avalonia doesn't have APIs for windows yet.

Does this mean this feature is on its way? If so, i'll just park this for now.

When using the Screens API, I noticed that `Screens.All` properly returns all displays. So there's that.

> if you need to render Avalonia inside of custom activities/presentations. But we don't have samples for that.

This is exactly what I tried to figure out. The point would be to have a shared UserControl that works across all platforms.

I attempted to use `await DrawingContextHelper.RenderAsync(...);` and whatnot, without much luck.

If a sample could be created for this, i'd be able to figure this out on my own.

**Comment by DoubleDBE** [comment, marked as answer]
Solved it by using `AvaloniaView` and setting the Content to the UserControl in question. Works fine!

```
        void IRecipient<CustomerDisplayRequest>.Receive(CustomerDisplayRequest message)
        {
            DisplayManager displayManager = (DisplayManager)GetSystemService(Context.DisplayService);

            Display[] presentationDisplays = displayManager.GetDisplays(DisplayManager.DisplayCategoryPresentation);

            if (presentationDisplays?.Length > 0)
            {
                Display display = presentationDisplays[0];

                var presentation = new Presentation(this, display);

                var avView = new AvaloniaView(this)
                {
                    Content = message.UserControl
                };

                presentation.SetContentView(avView);

                presentation.Show();
            }
        }
```

**Reply by Duncan-Firehouse** [reply]
Where is DisplayManager coming from? Is this using a Mono Android package or something? Also, where in the Avalonia solution would this code live?
